token-tracker/src/truth-sources/lp-tokens.js

```diff
@@ -0,0 +1,267 @@
+      KEEPETH_PAIR
+    )
+    const lpRewardKeepEth = new Contract(
+      LPRewardsKEEPETHJson,
+      this.context.web3
+    )
+
+    const KEEPTBTCPairTokenAbi = JSON.parse(keepTbtcTokenJson.result)
+    const keepTbtcTokenContract = new this.context.web3.eth.Contract(
+      KEEPTBTCPairTokenAbi,
+      KEEPTBTC_PAIR
+    )
+    const lpRewardKeepTbtc = new Contract(
+      LPRewardsKEEPTBTCJson,
+      this.context.web3
+    )
+
+    this.liquidityStakingObjects = {
+      KEEPETH: {
+        lpTokenContract: keepEthTokenContract,
+        lpRewardsContract: await lpRewardKeepEth.deployed(),
+        lpCreationBlock: KEEPETH_CREATION_BLOCK,
+        keepInLpTokenFilePath: KEEP_IN_LP_KEEPETH_BALANCES_PATH,
+        lpPairAddress: KEEPETH_PAIR,
+      },
+      KEEPTBTC: {
+        lpTokenContract: keepTbtcTokenContract,
+        lpRewardsContract: await lpRewardKeepTbtc.deployed(),
+        lpCreationBlock: KEEPTBTC_CREATION_BLOCK,
+        keepInLpTokenFilePath: KEEP_IN_LP_KEEPTBTC_BALANCES_PATH,
+        lpPairAddress: KEEPTBTC_PAIR,
+      },
+    }
+  }
+
+  /**
+   * Finds all historic stakers of LP KEEP-ETH / KEEP-TBTC pair token based on
+   * "Transfer" events
+   *
+   * @param {String} pairName LP pair name
+   *
+   * @return {Set<Address>} All historic LP KEEP-ETH / KEEP-TBTC token stakers
+   * */
+  async findStakers(pairName) {
+    const lpRewardsContractAddress = this.liquidityStaking.lpRewardsContract
+      .options.address
+
+    logger.info(
+      `looking for Transfer events emitted from ${lpRewardsContractAddress} ` +
+        `to ${pairName} pair ${this.liquidityStaking.lpTokenContract.options.address} ` +
+        `between blocks ${this.liquidityStaking.lpCreationBlock} and ${this.targetBlock}`
+    )
+
+    const events = await getPastEvents(
+      this.context.web3,
+      this.liquidityStaking.lpTokenContract,
+      "Transfer",
+      this.liquidityStaking.lpCreationBlock,
+      this.targetBlock
+    )
+    logger.info(`found ${events.length} lp ${pairName} token transfer events`)
+
+    const lpTokenStakersSet = new Set()
+    events.forEach((event) => {
+      // include accounts that staked in LPReward contract only
+      if (event.returnValues.to == lpRewardsContractAddress) {
+        lpTokenStakersSet.add(event.returnValues.from)
+      }
+    })
+
+    logger.info(`found ${lpTokenStakersSet.size} unique historic stakers`)
+
+    return Array.from(lpTokenStakersSet)
+  }
+
+  /**
+   * Retrieves balances of LP KEEP-ETH / KEEP-TBTC pair for stakers in LPRewards* contract
+   *
+   * @param {Array<Address>} lpStakers LP KEEP-ETH / KEEP-TBTC stakers
+   *
+   * @return {Map<Address,BN>} LP Balances by lp stakers
+   */
+  async getLpTokenStakersBalances(lpStakers) {
+    const lpBalanceByStaker = new Map()
+    let expectedTotalSupply = new BN(0)
+
+    for (let i = 0; i < lpStakers.length; i++) {
+      const lpBalance = new BN(
+        await this.liquidityStaking.lpRewardsContract.methods
```

Comment:
Do you think we should use `callWithRetry` everywhere when calling a method of a contract? And what is the reason in this case for calling with `callWithRetry`? Is it because of some `alchemy` issues?

Reply:
Yes, I believe so we should use it everywhere. It's unrelated to the provider, in case of any call failure we will retry. Better be safe than sorry 😉

Reply:
4db6e00 okay, I just wasn't sure when exactly we can experience that. Anyway, the only thing that bothers me now in  `callWithRetry` are the `making attempt number 0` in the logs. Especially it clutters things when retrieving ex. balances in a loop. Any idea how to not show it?

Reply:
We got rid of them in https://github.com/keep-network/keep-core/pull/2350, commit https://github.com/keep-network/keep-core/pull/2350/commits/2cb014758668626668693a2937c377bf48269acb so I am surprised they are still showing up. I think this branch is not fully merged with `master`.

Reply:
ah, yes, I was behind master, it's gone now, all good